modules/ximgproc/src/scansegment.cpp

```diff
@@ -0,0 +1,772 @@
+        for (int i = range.start; i < range.end; i++) {
+            OP1(i);
+        }
+        });
+
+    if (merge) {
+        // get cutoff size for clusters
+        std::vector<std::pair<int, int>> countVec;
+        int clusterIndexSize = clusterIndex.load();
+        countVec.reserve(clusterIndexSize / 2);
+        for (int i = 1; i < clusterIndexSize; i += 2) {
+            int count = clusterBuffer.data()[i];
+            if (count >= smallClusters) {
+                int currentID = clusterBuffer.data()[i - 1];
+                countVec.push_back(std::make_pair(currentID, count));
+            }
+        }
+
+        // sort descending
+        std::sort(countVec.begin(), countVec.end(), [](const std::pair<int, int>& left, const std::pair<int, int>& right) {
+            return left.second > right.second;
+            });
+
+        int countSize = (int)countVec.size();
+        int cutoff = MAX(smallClusters, countVec[MIN(countSize - 1, superpixels - 1)].second);
+        clusterCount = (int)std::count_if(countVec.begin(), countVec.end(), [&cutoff](std::pair<int, int> p) {return p.second > cutoff; });
+
+        // change labels to 1 -> clusterCount, 0 = UNKNOWN, reuse clusterbuffer
+        std::fill_n(clusterBuffer.data(), indexSize, UNKNOWN);
+        int countLimit = cutoff == -1 ? (int)countVec.size() : clusterCount;
+        for (int i = 0; i < countLimit; i++) {
+            clusterBuffer.data()[countVec[i].first] = i + 1;
+        }
+
+        parallel_for_(Range(0, (int)indexProcessVec.size()), [&](const Range& range) {
+            for (int i = range.start; i < range.end; i++) {
+                OP2(indexProcessVec[i]);
+            }
+            });
+
+        // make copy of labels buffer
+        memcpy(labelsMat.data, labelsBuffer.data(), indexSize * sizeof(int));
+
+        // run watershed
+        cv::parallel_for_(Range(0, (int)indexNeighbourVec.size()), [&](const Range& range) {
+            for (int i = range.start; i < range.end; i++) {
+                OP3(i);
+            }
+            });
+
+        // copy back to labels mat
+        parallel_for_(Range(0, (int)indexProcessVec.size()), [&](const Range& range) {
+            for (int i = range.start; i < range.end; i++) {
+                OP4(indexProcessVec[i]);
+            }
+            });
+    }
+    else {
+        memcpy(labelsMat.data, labelsBuffer.data(), indexSize * sizeof(int));
+    }
+
+    src.release();
+}
+
+void ScanSegmentImpl::OP1(int v)
+{
+    cv::Rect seedRect = seedRects.data()[v];
+    for (int y = seedRect.y; y < seedRect.y + seedRect.height; y++) {
+        for (int x = seedRect.x; x < seedRect.x + seedRect.width; x++) {
+            expandCluster(offsetVec[v].data(), cv::Point(x, y));
+        }
+    }
+}
+
+void ScanSegmentImpl::OP2(std::pair<int, int> const& p)
+{
+    std::pair<int, int>& q = const_cast<std::pair<int, int>&>(p);
+    for (int i = q.first; i < q.second; i++) {
```

Comment:
Why do we need this alias?

> `std::pair<int, int>& q = const_cast<std::pair<int, int>&>(p);`


---

The same note is about OP4

Reply:
My apologies. This was some old code that got carried over. Have updated it.